embassy-net-adin1110/src/lib.rs

```diff
@@ -0,0 +1,1249 @@
+#![allow(clippy::pedantic)]
+#![feature(type_alias_impl_trait)]
+#![feature(async_fn_in_trait)]
+#![cfg_attr(not(any(test, feature = "std")), no_std)]
+
+use ch::driver::LinkState;
+use embassy_futures::select::{select, Either};
+use embassy_net_driver_channel as ch;
+use embassy_time::{Duration, Timer};
+use embedded_hal_1::digital::OutputPin;
+use embedded_hal_async::digital::Wait;
+use embedded_hal_async::spi::{Operation, SpiDevice};
+use heapless::Vec;
+
+mod crc32;
+mod crc8;
+mod mdio;
+mod phy;
+mod regs;
+
+pub use crc32::ETH_FSC;
+use crc8::crc8;
+pub use mdio::MdioBus;
+pub use phy::{Phy10BaseT1x, RegsC22, RegsC45};
+pub use regs::{Config0, Config2, SpiRegisters as sr, Status0, Status1};
+
+use crate::regs::{LedCntrl, LedFunc, LedPol, LedPolarity, SpiHeader};
+
+pub const PHYID: u32 = 0x0283BC91;
+
+/// Error values ADIN1110
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[allow(non_camel_case_types)]
+pub enum AdinError<E> {
+    Spi(E),
+    SENDERROR,
+    READERROR,
+    CRC,
+    PACKET_TOO_BIG,
+    PACKET_TOO_SMALL,
+    MDIO_ACC_TIMEOUT,
+}
+
+pub type AEResult<T, SPIE> = core::result::Result<T, AdinError<SPIE>>;
+pub const MDIO_PHY_ADDR: u8 = 0x01;
+pub const MTU: usize = 1500;
+/// Max SPI/Frame buffer size
+pub const MAX_BUFF: usize = 2048;
+
+const DONT_CARE_BYTE: u8 = 0x00;
+const TURN_AROUND_BYTE: u8 = 0x00;
+
+const FEC_LEN: usize = 4;
+const FRAME_HEADER_LEN: usize = 2;
+const WR_HEADER_LEN: usize = 2;
+
+// P1 = 0x00, P2 = 0x01
+const PORT_ID_BYTE: u8 = 0x00;
+
+pub type Packet = Vec<u8, { MTU + FEC_LEN + WR_HEADER_LEN }>;
+
+/// Type alias for the embassy-net driver for ADIN1110
+pub type Device<'d> = embassy_net_driver_channel::Device<'d, MTU>;
+
+/// Internal state for the embassy-net integration.
+pub struct State<const N_RX: usize, const N_TX: usize> {
+    ch_state: ch::State<MTU, N_RX, N_TX>,
+}
+impl<const N_RX: usize, const N_TX: usize> State<N_RX, N_TX> {
+    /// Create a new `State`.
+    pub const fn new() -> Self {
+        Self {
+            ch_state: ch::State::new(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ADIN1110<SPI> {
+    /// SPI bus
+    spi: SPI,
+    /// Enable CRC on SPI transfer.
+    /// This must match with the hardware pin SPI_CFG0 were 0 = CRC enable, 1 CRC disabled.
+    crc: bool,
+}
+
+// Round size up the N u32;
+pub(crate) fn size_align_u32(size: u32) -> u32 {
+    (size + 3) & 0xFFFF_FFFC
+}
+
+impl<SpiE, SPI> ADIN1110<SPI>
```

Comment:
to avoid adding the error as an extra generic param, you can do:

```rust
impl< SPI> ADIN1110<SPI>
where
    SPI: SpiDevice<u8>,
```

and then refer to the error as `SPI::Error`.

Reply:
I refactored the code so it is not a generic parameter anymore.